Compression Streams not handling large outputs during the flush stage

stwrt opened the pull request:
#### a30d9284f5b98bd5c389f28bfc4a710d9f256fa8
<pre>
Compression Streams not handling large outputs during the flush stage
<a href="https://bugs.webkit.org/show_bug.cgi?id=254021">https://bugs.webkit.org/show_bug.cgi?id=254021</a>

Reviewed by Chris Dumez and Brent Fulgham.

We missed an edge case where during the flush step we may have data longer than
the allocated output. Since the avail_in will be set to 0 we would just exit.
We need to verify that the stream has ended before exiting.

* LayoutTests/imported/w3c/web-platform-tests/compression/compression-large-flush-output.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-large-flush-output.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-large-flush-output.any.js: Added.
(async compressData):
(promise_test.async t):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
(WebCore::CompressionStreamEncoder::flush):
(WebCore::CompressionStreamEncoder::didDeflateFinish const):
(WebCore::didDeflateFail):
(WebCore::CompressionStreamEncoder::compress):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
(WebCore::CompressionStreamEncoder::CompressionStreamEncoder):
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
(WebCore::DecompressionStreamDecoder::flush):
(WebCore::DecompressionStreamDecoder::didInflateFinish const):
(WebCore::didInflateFail):
(WebCore::DecompressionStreamDecoder::didInflateContainExtraBytes const):
(WebCore::DecompressionStreamDecoder::decompressZlib):
(WebCore::DecompressionStreamDecoder::decompressAppleCompressionFramework):
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
(WebCore::DecompressionStreamDecoder::DecompressionStreamDecoder):

Canonical link: <a href="https://commits.webkit.org/263997@main">https://commits.webkit.org/263997@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ab429a7b4d3454bbfbc4c47e2856d499c1a9fb7a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6353 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6551 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6733 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7923 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6658 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6348 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6770 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6500 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9554 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6465 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6466 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7997 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3927 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5729 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5794 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5809 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6290 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-WK2-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5697 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/5657 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1502 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9855 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6067 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -133,7 +132,7 @@ ExceptionOr<RefPtr<JSC::ArrayBuffer>> CompressionStreamEncoder::compress(const u
             return Exception { TypeError, "Failed to compress data."_s };
 
-        if (!m_zstream.avail_in) {
+        if (!m_zstream.avail_in && (!m_finish || (m_finish && result == Z_STREAM_END))) {
```
`m_finish` is badly named. Boolean variables in WebKit need a prefix as per coding style. Should probably be something like `m_didFinish` or `m_isFinished`.

brentfulgham replied:
`m_finish` is existing code, so renaming would increase the size of this bug fix. But it certainly seems like a good change to make.

I also feel like this conditional is confusing to read/understand. I think it's saying that if we have consumed all the input data, and are either not finished, or finished but didn't have enough output buffer space, we should expand the output buffer and continue.

I suggest you make a helper function with a name that explains this, and call it here (and in the similar DecompressionStream code) because the Zlib API is very terse and not easy to follow.

brentfulgham commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -133,7 +132,7 @@ ExceptionOr<RefPtr<JSC::ArrayBuffer>> CompressionStreamEncoder::compress(const u
         if (result != Z_OK && result != Z_STREAM_END && result != Z_BUF_ERROR)
```
Nit: result is scoped to the body of this loop. I think it's clearer to declare it on the line above as it is created, rather than defining it externally without initializing.

brentfulgham reviewed (changes requested):
I think this looks like it fixes the issue, but I think the code would be better with some renaming to clarify what's going on.

I'd also like to see a test case so we can avoid this breaking in the future.

brentfulgham reviewed (changes requested):
I think this looks like it fixes the issue, but I think the code would be better with some renaming to clarify what's going on.

I'd also like to see a test case so we can avoid this breaking in the future.

brentfulgham reviewed (commented):
This looks much better! Thank you for creating test cases for this.

brentfulgham reviewed (approved):
Looks good!

cdumez commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
```diff
@@ -60,11 +60,14 @@ class CompressionStreamEncoder : public RefCounted<CompressionStreamEncoder> {
 
     Formats::CompressionFormat m_format;
 
+    bool didDeflateFinish(int);
```
Member functions should come *before* data members, not in the middle of them.

cdumez replied:
Also, can this be marked as `const`?

cdumez commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
```diff
@@ -88,6 +88,10 @@ class DecompressionStreamDecoder : public RefCounted<DecompressionStreamDecoder>
 
     Formats::CompressionFormat m_format;
 
+    bool didInflateFinish(int);
```
Same comments apply for these here.

cdumez commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
```diff
@@ -200,13 +221,13 @@ ExceptionOr<RefPtr<JSC::ArrayBuffer>> DecompressionStreamDecoder::decompressAppl
         m_stream.dst_size = output.size();
 
-        result = compression_stream_process(&m_stream, (m_finish) ? COMPRESSION_STREAM_FINALIZE : 0);
+        result = compression_stream_process(&m_stream, (m_didFinish) ? COMPRESSION_STREAM_FINALIZE : 0);
```
No need for `()` around `m_didFinish`.

cdumez replied:
Still extra parentheses, this wasn't resolved.

cdumez commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -129,11 +145,12 @@ ExceptionOr<RefPtr<JSC::ArrayBuffer>> CompressionStreamEncoder::compress(const u
 
-        result = deflate(&m_zstream, (m_finish) ? Z_FINISH : Z_NO_FLUSH);
-        if (result != Z_OK && result != Z_STREAM_END && result != Z_BUF_ERROR)
+        result = deflate(&m_zstream, (m_didFinish) ? Z_FINISH : Z_NO_FLUSH);
```
No need for `()` around `m_didFinish`

cdumez commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -95,9 +95,25 @@ ExceptionOr<bool> CompressionStreamEncoder::initialize()
+}
+
+// See https://www.zlib.net/manual.html#Constants
+bool CompressionStreamEncoder::didDeflateFail(int result) const
```
This doesn't use any data members, it can be static.

cdumez commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
```diff
@@ -95,6 +95,28 @@ ExceptionOr<bool> DecompressionStreamDecoder::initialize()
+}
+
+// See https://www.zlib.net/manual.html#Constants
+bool DecompressionStreamDecoder::didInflateFail(int result) const
```
This doesn't use any data members, it can be static.

stwrt committed:
Compression Streams not handling large outputs during the flush stage
https://bugs.webkit.org/show_bug.cgi?id=254021

Reviewed by Chris Dumez and Brent Fulgham.

We missed an edge case where during the flush step we may have data longer than
the allocated output. Since the avail_in will be set to 0 we would just exit.
We need to verify that the stream has ended before exiting.

* LayoutTests/imported/w3c/web-platform-tests/compression/compression-large-flush-output.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-large-flush-output.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/compression/compression-large-flush-output.any.js: Added.
(async compressData):
(promise_test.async t):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
(WebCore::CompressionStreamEncoder::flush):
(WebCore::CompressionStreamEncoder::didDeflateFinish const):
(WebCore::didDeflateFail):
(WebCore::CompressionStreamEncoder::compress):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
(WebCore::CompressionStreamEncoder::CompressionStreamEncoder):
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
(WebCore::DecompressionStreamDecoder::flush):
(WebCore::DecompressionStreamDecoder::didInflateFinish const):
(WebCore::didInflateFail):
(WebCore::DecompressionStreamDecoder::didInflateContainExtraBytes const):
(WebCore::DecompressionStreamDecoder::decompressZlib):
(WebCore::DecompressionStreamDecoder::decompressAppleCompressionFramework):
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
(WebCore::DecompressionStreamDecoder::DecompressionStreamDecoder):

Canonical link: https://commits.webkit.org/263997@main
